src/tools/ceph_dedup/ceph_dedup_daemon.cc

```diff
@@ -0,0 +1,817 @@
+	  begin,
+	  end,
+	  i,
+	  max_thread,
+	  &shard_start,
+	  &shard_end);
+
+	threads.emplace_back(
+	  io_ctx,
+	  chunk_io_ctx,
+	  shard_start,
+	  shard_end,
+	  chunk_size,
+	  fp_algo,
+	  chunk_algo,
+	  sample_dedup_global,
+	  snap);
+	threads.back().create("sample_dedup");
+      }
+    }
+
+    for (auto &p : threads) {
+      p.join();
+      total_size += p.get_total_object_size();
+      total_duplicate_size += p.get_total_duplicated_size();
+    }
+
+    cerr << "Summary: read "
+	 << total_size << " bytes so far and found saveable space ("
+	 << total_duplicate_size << " bytes)."
+	 << std::endl;
+
+    {
+      lock_guard lock(glock);
+      threads.clear();
+      if (run_once) {
+	all_stop = true;
+	break;
+      }
+    }
+    sleep(wakeup_period);
+
+    map<string, librados::pool_stat_t> stats;
+    ret = rados.get_pool_stats(pool_names, stats);
+    if (ret < 0) {
+      cerr << "error fetching pool stats: " << cpp_strerror(ret) << std::endl;
+      return -EINVAL;
+    }
+    if (stats.find(base_pool_name) == stats.end()) {
+      cerr << "stats can not find pool name: " << base_pool_name << std::endl;
+      return -EINVAL;
+    }
+  }
+
+  return 0;
+}
+
+static void handle_signal(int signum) 
+{
+  std::lock_guard l{glock};
+  all_stop = true;
+  for (auto &p : threads) {
+    p.signal(signum);
+  }
+}
+
+int main(int argc, const char **argv)
+{
+  auto args = argv_to_vec(argc, argv);
+  if (args.empty()) {
+    cerr << argv[0] << ": -h or --help for usage" << std::endl;
+    exit(1);
+  }
+
+  po::variables_map opts;
+  po::positional_options_description p;
+  p.add("command", 1);
+  po::options_description desc = make_usage();
+  try {
+    po::parsed_options parsed =
+      po::command_line_parser(argc, argv).options(desc).positional(p).allow_unregistered().run();
+    po::store(parsed, opts);
+    po::notify(opts);
+  } catch(po::error &e) {
+    std::cerr << e.what() << std::endl;
+    return 1;
+  }
+  if (opts.count("help") || opts.count("h")) {
+    cout<< desc << std::endl;
+    exit(0);
+  }
+
+  auto cct = global_init(NULL, args, CEPH_ENTITY_TYPE_CLIENT,
+			CODE_ENVIRONMENT_DAEMON,
+			CINIT_FLAG_UNPRIVILEGED_DAEMON_DEFAULTS);
+
+  Preforker forker;
+  if (global_init_prefork(g_ceph_context) >= 0) {
+    std::string err;
+    int r = forker.prefork(err);
+    if (r < 0) {
+      cerr << err << std::endl;
+      return r;
+    }
+    if (forker.is_parent()) {
+      g_ceph_context->_log->start();
+      if (forker.parent_wait(err) != 0) {
+        return -ENXIO;
+      }
+      return 0;
+    }
+    global_init_postfork_start(g_ceph_context);
+  }
+  common_init_finish(g_ceph_context);
+  if (g_conf()->daemonize) {
+    global_init_postfork_finish(g_ceph_context);
+    forker.daemonize();
+  }
+
+  init_async_signal_handler();
+  register_async_signal_handler_oneshot(SIGINT, handle_signal);
+  register_async_signal_handler_oneshot(SIGTERM, handle_signal);
+
+  int ret = make_crawling_daemon(opts);
```

Comment:
Before your next work item, submit a PR renaming make_crawling_daemon to run_crawling_daemon.

Reply:
OK.